girder_annotation/girder_large_image_annotation/rest/annotation.py

```diff
@@ -558,6 +562,142 @@ def deleteItemAnnotations(self, item):
+    @access.public
+    def existFolderAnnotations(self, id, recurse):
+        user = self.getCurrentUser()
+        if not user:
+            return []
+        annotations = self.getFolderAnnotations(id, recurse, user, 1)
+        try:
+            next(annotations)
+            return True
+        except StopIteration:
+            return False
+
+    @autoDescribeRoute(
+        Description('Get the user-owned annotations from the items in a folder')
+        .param('id', 'The ID of the folder', required=True, paramType='path')
+        .param('recurse', 'Whether or not to retrieve all '
+               'annotations from subfolders', required=False, default=False, dataType='boolean')
+        .pagingParams(defaultSort='created', defaultSortDir=-1)
+        .errorResponse()
+    )
+    @access.public
+    def returnFolderAnnotations(self, id, recurse, limit, offset, sort):
+        user = self.getCurrentUser()
+        if not user:
+            return []
+        annotations = self.getFolderAnnotations(id, recurse, user, limit, offset,
+                                                sort[0][0], sort[0][1])
+
+        def count():
+            try:
+                return next(self.getFolderAnnotations(id, recurse, self.getCurrentUser(),
+                                                      count=True))['count']
+            except StopIteration:
+                # If there are no values to iterate over, the count is 0 and should be returned
+                return 0
+
+        annotations.count = count
+        return annotations
+
+    @autoDescribeRoute(
+        Description('Set the access for all the user-owned annotations from the items in a folder')
+        .param('id', 'The ID of the folder', required=True, paramType='path')
+        .param('access', 'The JSON-encoded access control list.')
+        .param('public', 'Whether the annotation should be publicly visible.',
+               dataType='boolean', required=False)
+        .param('recurse', 'Whether or not to retrieve all '
+               'annotations from subfolders', required=False, default=False, dataType='boolean')
+        .errorResponse('ID was invalid.')
+    )
+    @access.user
+    def setFolderAnnotationAccess(self, id, params):
+        user = self.getCurrentUser()
+        if not user:
+            return []
+        access = json.loads(params['access'])
+        public = self.boolParam('public', params, False)
+        for annotation in self.getFolderAnnotations(id, params['recurse'], user):
+            annot = Annotation().load(annotation['_id'], user=user, getElements=False)
+            annot = Annotation().setPublic(annot, public)
+            annot = Annotation().setAccessList(
+                annot, access, save=False, user=user)
+            Annotation().update({'_id': annot['_id']}, {'$set': {
+                key: annot[key] for key in ('access', 'public', 'publicFlags')
+                if key in annot
+            }})
+            yield annot
```

Comment:
We might be change a *lot* of annotations, so returning them all is probably not desired.  We may just want to return a summary (maybe `{'updated': <count>}`)?

Also, since this can take a while to run, we should add `setResponseTimeLimit()` inside the loop.

Reply:
That makes sense.

See 666ee38